Implement ChainFragment as a Finger Tree

mrBliss opened the pull request:
A `ChainFragment` represents a (possibly empty) fragment of the chain. It is
based on a finger tree for efficient searching based on the `Slot` or `Point`
of a block.

This implementation is based on `archive/FingerTree.hs`. It has been updated
to (better) adhere to the API of `Ouroboros.Network.Chain`, the reference
implementation based on a snoc list.

We need to come up with a better way to work with empty `ChainFragment`s. At
the moment, to mimic `Ouroboros.Network.Chain`, an empty `ChainFragment` is
treated more or less as a Genesis block.

The test for `lookupByIndexFromEnd` is still failing, I have to fix this.

mrBliss commented:
Worth mentioning: in 8d6fd38bc2bacecb1fb47824e996c88449e0ce2a I discovered a bug in the implementation of `search`, which is part of the`fingertree` library. The `search` function is a newer function, which was not part of the original paper on Finger Trees. It also didn't have any tests. I filed a [bug report](https://hub.darcs.net/ross/fingertree/issue/8) with a fix and tests.

Because of the bug, I rewrote some functions that were using `search` in terms of the correct `split`function.

Ross swiftly accepted the fix and published a new version on Hackage: `fingertree-0.1.4.2`.

In c0df49e947b35544eb299b23043735cd66a104c0 (and also 1cffcc236e7d913aec1bcb8fada0fdfcc5f6937c) I bumped the dependency and restored the original implementations I had that were based on `search`.

adinapoli-iohk commented:
Great catch on the fingertree library, @mrBliss !

edsko commented on src/Ouroboros/Network/ChainFragment.hs:
```diff
@@ -0,0 +1,540 @@
+    c = ChainFragment $ FT.fromList bs
+
+-- | Drop the last @n@ blocks from the 'ChainFragment'.
+drop :: HasHeader block
```
This is perhaps confusingly named. `drop` normally drops from the _head_ of the list, but here it's dropping from the end. Might be good to pick a more descriptive name. `dropNewest` perhaps?

mrBliss replied:
Good suggestion, I'll rename it to `dropNewest`.

edsko commented on src/Ouroboros/Network/ChainFragment.hs:
```diff
@@ -0,0 +1,540 @@
+
+
+length :: ChainFragment block -> Int
+length (ChainFragment c) = Foldable.length c
```
Is there a reason _not_ to declare `ChainFragment` itself to be foldable? Then all these wrappers around `Foldable.stuff` would not be necessary right?

mrBliss replied:
My reasoning is that the exposed wrappers are explicit about the ordering (right-to-left or the other way around, or oldest first, etc). Just exposing the `Foldable` instance would shift the burden of thinking about this to the users.

But I don't mind adding a `Foldable (ChainFragment block)` instance if you like.

edsko replied:
Aha, gotcha. Yup, sounds reasonable.

mrBliss commented on src/Ouroboros/Network/Block.hs:
```diff
@@ -47,6 +50,11 @@ class StandardHash b => HasHeader b where
+-- | When implementing 'HasHeader', use this method to implement the 'measure'
+-- method of the 'Measured' super class.
+blockMeasure :: HasHeader b => b -> BlockMeasure
+blockMeasure b = BlockMeasure (blockSlot b) (blockSlot b) 1
```
The point of this function is the following: all the implementers of `HasHeader` have to implement the `Measured` type class. Instead of always writing the following:

```haskell
instance Measured BlockMeasure SomeBlockType where
  measure b = BlockMeasure (blockSlot b) (blockSlot b) 1
```

They can write:

```haskell
instance Measured BlockMeasure SomeBlockType where
  measure = blockMeasure
```

So it is just a convenience function that 'makes sure' the write measure is used.

edsko replied:
I see what you mean. My point above still stands though... `HasHeader` _implies_ `Measured`... :thinking:

mrBliss replied:
True, but you can use a function that uses the `HasHeader` instance to implement its `Measured` super-class instance, as long as you are not using the `Measured` super-class instance in that function, which is the case here.

dcoutts commented on src/Ouroboros/Network/ChainFragment.hs:
```diff
@@ -0,0 +1,536 @@
+
+-- | Make a list of blocks from a 'ChainFragment', in newest-to-oldest order.
+toNewestFirst :: HasHeader block => ChainFragment block -> [block]
+toNewestFirst = foldChainFragment (flip (:)) []
```
Is this O(n log n)?

Note how the fingertree `Foldable` instance works. They do not provide a special `toList` operation, working instead with `foldMap` that goes over the structure of the tree. It reles on the default `toList` in terms of `foldMap`. So we're using 'foldChainFragment' here which uses `Foldable.foldl'`. It's worth checking what complexity this would give us and if we can do at least as well as `Foldable.toList` does for `FingerTree` which the docs claim is O(n).

mrBliss commented on src/Ouroboros/Network/ChainFragment.hs:
```diff
@@ -0,0 +1,540 @@
+--
+-- This function does not check whether any of the two blocks satisfy
+-- 'blockInvariant'.
+isValidSuccessorOf :: HasHeader block
```
Yes, that makes sense. I was expecting this function to already exist, actually.

mrBliss replied:
I just noticed when trying to move this function over to `Chain` and using it for `Chain.validExtension` that it doesn't fit well with Genesis and `Point` vs. `Block`.

mrBliss commented on src/Ouroboros/Network/ChainFragment.hs:
```diff
@@ -0,0 +1,540 @@
+head (_ :> b) = Just b
+head Empty    = Nothing
+
+headPoint :: HasHeader block => ChainFragment block -> Maybe (Point block)
```
No reason, I'll change it

mrBliss commented on src/Ouroboros/Network/Testing/ConcreteBlock.hs:
```diff
@@ -154,6 +165,42 @@ fixupBlockHeader c h b = b'
+-------------------------------------------------------------------------------}
+
+-- TODO avoid code duplication by a ChainLike type class? Or just by
+-- parameterising over the 'headHash' and the 'headBlockNo' functions?
```
That's what I was thinking. I'll remove this TODO.

mrBliss commented on test/Test/ChainFragment.hs:
```diff
@@ -0,0 +1,737 @@
+
+prop_lookupBySlot :: TestChainFragmentAndPoint -> Property
+prop_lookupBySlot (TestChainFragmentAndPoint c p) =
+  CF.pointOnChainFragment p c ==>
```
The issue here is that `pointOnChainFragment` checks for a matching `Point` (slot + hash) whereas `lookupBySlot` only looks at the slot. A solution would be to add `slotOnChainFragment` and to use it in the Nothing case here.

mrBliss commented on test/Test/ChainFragment.hs:
```diff
@@ -0,0 +1,737 @@
+
+prop_dropOldest_Empty :: TestBlockChainFragment -> Bool
+prop_dropOldest_Empty (TestBlockChainFragment chain) =
+    CF.drop (CF.length chain) chain == Empty
```
Indeed!

mrBliss commented on test/Test/ChainFragment.hs:
```diff
@@ -0,0 +1,737 @@
+prop_splitBeforeSlot_pointOnChainFragment (TestChainFragmentAndPoint c p) =
+  case CF.splitBeforeSlot c (pointSlot p) of
+    Just (l, r) -> not (CF.pointOnChainFragment p l)
+                && CF.joinChainFragments l r == Just c
```
Still, even though it's partial, the join should result in a `Just c`. I have applied your suggestion about comparing the joins instead of the sum of the lengths.

dcoutts commented on src/Ouroboros/Network/ChainFragment.hs:
```diff
@@ -0,0 +1,561 @@
+-- >                          └───┘     ├───┤
+-- >                                    │ 8B│
+-- >                                    └───┘
+-- > Just (l1,       l2,        r1,       r2)
```
:+1: Very nice.

dcoutts commented on src/Ouroboros/Network/ChainFragment.hs:
```diff
@@ -0,0 +1,561 @@
+               -- splitAfterPoint initC2 p cannot fail, since p comes out of
+               -- initC2
+               -> Just (l1, l2, r1, r2)
+             _ -> go c1 c2'
```
So just to check I've got it. This is O(|r2| * log |c1|) right? Since we're going down c2 block by block and looking for the point in the corresponding chain.

I'll have a think about whether we can do something that's log time in both sides.

mrBliss commented on src/Ouroboros/Network/ChainFragment.hs:
```diff
@@ -0,0 +1,561 @@
+length (ChainFragment c) = Foldable.length c
+
+null :: ChainFragment block -> Bool
+null (ChainFragment c) = Foldable.null c
```
Good catch. I have re-implemented `length` using the measure to obtain a O(1) complexity. I have also re-implemented `null` in terms of `FT.null`, which is documented to have O(1), instead of relying on the undocumented complexity of `Foldable.null`.

mrBliss commented:
I have annotated most functions with their time complexity. I am using the markup used in [`Data.Sequence`](http://hackage.haskell.org/package/containers-0.6.0.1/docs/Data-Sequence.html) for this, e.g. `\( O(n) \)`.

P.S. I have addressed the comment about `length` not being O(1), but the "Resolve Conversation" button is missing.

mrBliss commented:
I'll rebase it against master and squash the reviewed commits. After the review, I'll squash the remainder of the commits.

mrBliss commented:
> I'll rebase it against master and squash the reviewed commits. After the review, I'll squash the remainder of the commits.

I'll do it Monday morning.
